node/src/store/in_memory_store.rs

```diff
@@ -1,199 +1,306 @@
-        self.get_by_height(head_height)
+    /// Clone the store and all its contents. Async fn due to internal use of async mutex.
+    pub async fn async_clone(&self) -> Self {
+        InMemoryStore {
+            inner: RwLock::new(self.inner.read().await.clone()),
+            sampling_data: self.sampling_data.clone(),
+            header_added_notifier: Notify::new(),
+        }
+    }
+}
+
+impl InMemoryStoreInner {
+    fn get_stored_ranges(&self) -> HeaderRanges {
+        self.stored_ranges.clone()
+    }
+
+    #[inline]
+    fn get_head_height(&self) -> Result<u64> {
+        self.stored_ranges.head().ok_or(StoreError::NotFound)
     }
 
     fn contains_hash(&self, hash: &Hash) -> bool {
         self.headers.contains_key(hash)
     }
 
     fn get_by_hash(&self, hash: &Hash) -> Result<ExtendedHeader> {
-        self.headers
-            .get(hash)
-            .as_deref()
-            .cloned()
-            .ok_or(StoreError::NotFound)
+        self.headers.get(hash).cloned().ok_or(StoreError::NotFound)
     }
 
     fn contains_height(&self, height: u64) -> bool {
-        let Ok(head_height) = self.get_head_height() else {
-            return false;
-        };
-
-        height != 0 && height <= head_height
+        self.stored_ranges.contains(height)
     }
 
     fn get_by_height(&self, height: u64) -> Result<ExtendedHeader> {
-        if !self.contains_height(height) {
+        let Some(hash) = self.height_to_hash.get(&height).copied() else {
             return Err(StoreError::NotFound);
-        }
-
-        let Some(hash) = self.height_to_hash.get(&height).as_deref().copied() else {
-            return Err(StoreError::LostHeight(height));
         };
 
         self.headers
             .get(&hash)
-            .as_deref()
             .cloned()
             .ok_or(StoreError::LostHash(hash))
     }
 
-    fn update_sampling_metadata(
-        &self,
-        height: u64,
-        status: SamplingStatus,
-        cids: Vec<Cid>,
+    async fn insert(
+        &mut self,
+        headers: Vec<ExtendedHeader>,
+        verify_neighbours: bool,
     ) -> Result<()> {
-        if !self.contains_height(height) {
-            return Err(StoreError::NotFound);
+        let (Some(head), Some(tail)) = (headers.first(), headers.last()) else {
+            return Ok(());
+        };
+
+        let headers_range = head.height().value()..=tail.height().value();
+        let range_scan_result = self.stored_ranges.check_range_insert(&headers_range)?;
+
+        if verify_neighbours {
+            let prev_exists = headers_range.start() != range_scan_result.range.start();
+            let next_exists = headers_range.end() != range_scan_result.range.end();
+            // header range is already internally verified against itself in `P2p::get_unverified_header_ranges`
+            self.verify_against_neighbours(head, tail, (prev_exists, next_exists))?;
+        } else {
+            verify_range_contiguous(&headers)?;
```

Comment:
Shouldn't we do this regardless? This issue exists in other stores too.

Reply:
Skipping this check here relied on it being run when validating the headers range. But similar to above, if we want to provide single "checked", interface this can go.